Add a cluster setting to disallow expensive queries

matriv opened the pull request:
Add a new cluster setting `search.allow_expensive_queries` which by
default is `true`. If set to `false`, certain queries that have 
usually slow performance cannot be executed and an error message
is returned.

- Queries that need to do linear scans to identify matches:
  - Script queries
- Queries that have a high up-front cost:
  - Fuzzy queries
  - Regexp queries
  - Prefix queries (without index_prefixes enabled
  - Wildcard queries
  - Range queries on text and keyword fields
- Joining queries
  - HasParent queries
  - HasChild queries
  - ParentId queries
  - Nested queries
- Queries on deprecated 6.x geo shapes (using PrefixTree implementation)
- Queries that may have a high per-document cost:
  - Script score queries
  - Percolate queries

Closes: #29050

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

matriv committed:
Add a cluster setting to disallow slow queries

Add a new cluster setting `search.disallow_slow_queries` which by
default is `false`. If set to `true` then certain queries (prefix,
fuzzy, regexp and wildcard) that have usually slow performance cannot
be executed and an exception is thrown.

Closes: #29050

jimczi commented on server/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -223,6 +226,8 @@ public IndexService(
         this.retentionLeaseSyncTask = new AsyncRetentionLeaseSyncTask(this);
         updateFsyncTaskIfNecessary();
+
+
```
nit: remove change

jimczi commented on docs/reference/query-dsl.asciidoc:
```diff
@@ -25,6 +25,14 @@ or to alter their behaviour (such as the
+<<query-dsl-regexp-query,`regexp`>> and <<query-dsl-bool-query,`wildcard`>> ,
+that are usually slow performance can affect the cluster performance.
+The execution of such queries can be prevented by setting the value of the `search.disallow_slow_queries`
+setting to `true` (defaults tp `false`).
```
nit: s/tp/to

jimczi reviewed (commented):
The new setting looks good, although I am still not decided whether we should add more granularity to the detection of slow queries. One possibility would be to have some static limits, like number of characters in the prefix query, wildcard with leading wildcards, ... Another possibility could be to hook into the rewrite of multi-terms to limit the expansion to few terms but the first option might be enough for most of the cases.
Should we also add the exception for `script` and `script_score` queries ? We could also try to restrict them only if they are the only required clause in the query. This would allow users to continue to use `script` but only if they have another required clause based on inverted lists.
I'd also like @jpountz and @romseygeek to take a look to validate the approach and the limitations that we want to enforce in this new setting.

matriv committed:
address comments

rjernst reviewed (commented):
One general suggestion: can we use positive logic instead of double negative? ie have the setting be `allow-slow-queries` with a default of `true`.

jasontedor commented:
Putting aside @rjernst's thoughts about the double negative, I have some general concerns about the name. The problem that I have is that this setting does not forbid slow queries (i.e., queries that from a user-facing perspective are taking a long period of time, which then get aborted if this setting is set) but rather queries that are "expensive" to execute, because they scale poorly. Perhaps the naming could reflect that, like `allow_expensive_queries`.

matriv commented:
I agree with both suggestions regarding the negative naming and the `expensive` vs `slow`.
@jimczi @jpountz @romseygeek what do you think?

matriv committed:
Add script and script_score queries to the disallowed ones

matriv commented:
Currently I have added integration tests for each one of the disallowed queries. For the Script/ScriptScore it's not possible to include them in the yml test since the required ScriptPlugins are not loaded for those tests, so instead I added a test method in the corresponding java `xxxIT` tests.

After some discussion I had with @rjernst, he suggested to just have one integ test, since the path of updating/passing the setting down to the QueryShardContext is the same for all the queries.

@jimczi @jpountz @romseygeek What's your opinion?

matriv committed:
Disallow joining queries

matriv commented:
@elasticmachine run elasticsearch-ci/2

matriv commented:
@rjernst @jasontedor Regarding the negative name and semantics, I have to add that the idea is as a next step to have more fine-grain control on the disallowed queries, so something like `search.disallow_expensive_queries : "fuzzy, regexp, wildcard"` which imo seems to promote the usage of a negative semantics setting. Otherwise we will have a setting with default value like this: `search.allow_expensive_queries : "fuzzy, joining, regexp, prefix, script, script_score, wildcard"` where one should start removing values from the setting.

rjernst commented:
I think allow/disallow should really be a boolean. If we want the user to control the exact queries allowed, it should be a different setting. I also think an inclusive setting, while more verbose, is much simpler for a user to understand and less error prone in the future. If the list setting has negative semantics, new queries which we deem expensive may automatically be included on upgrade when the user already decided which expensive queries to allow.

jpountz reviewed (commented):
Some thoughts about this PR:
 - Since I initially opened the issue that this PR addresses, the company seems to have embraced slow operations rather than prevented them (see e.g. searchable snapshots and the async search API) so I wonder whether this is still something we want to do. cc @giladgal
 - Regardless I think there is a theme around slow queries and we need to provide users with a better experience:
   - Make multi-term queries honor the `timeout` setting and task cancellation by leveraging `ExitableDirectoryReader`.
   - Better warn users when they run slow queries that have faster alternatives, such as running prefix queries on a text field that doesn't have `index_prefixes` enabled. There aren't many examples right now, but we expect several ones to come, e.g. when we enable querying numeric fields that have doc values but are not indexed, when we add scripted fields, or when the `wildcard_keyword` field comes out.

giladgal commented:
> * Since I initially opened the issue that this PR addresses, the company seems to have embraced slow operations rather than prevented them (see e.g. searchable snapshots and the async search API) so I wonder whether this is still something we want to do. cc @giladgal

I think the PR still makes sense. Giving users the tools to run resource-draining slow queries can make it easier for users to get themselves into troubles, which only makes such a setting more important as a safety mechanism.

matriv committed:
Rename setting to positive semantics

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

matriv committed:
Add percolator queries to the expensive ones

matriv committed:
Add legacy geo-shape queries to the disallowed ones

matriv committed:
capitalize the first letter of error messages

matriv committed:
Add range queries on text/keyword to the expensive ones

matriv committed:
Add index_prefixes suggestion to the error message

matriv committed:
fix error msg

matriv committed:
fix assertions of error messages

matriv committed:
Enhance part of allow expensive in docs Query DSL

matriv committed:
fix tests

matriv committed:
improve error messages

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

matriv committed:
fix test

imotov reviewed (commented):
I think it might make sense to add a test for BKD-based shapes as well to make sure they still work even if slow queries are disabled. Otherwise it looks reasonable, if we indeed want to disable these queries for legacy geo shapes.

matriv committed:
fix test

matriv committed:
fix test

matriv committed:
Check that allow_expensive_queries setting doesn't affect new geo-shape queries

matriv committed:
revert the legacy quad tree test to the same method

matriv committed:
remove unused import

jpountz commented on docs/reference/query-dsl.asciidoc:
```diff
@@ -25,6 +25,27 @@ or to alter their behaviour (such as the
+Those queries can be categorised as follows:
+* Queries that need to do linear scans to identify matches:
+** <<query-dsl-script-query, `script queries`>>
+** <<query-dsl-script-score-query, `script score queries`>>
```
I was expecting this one to be fine since it finds matches using another query?

matriv replied:
It was a suggestion by @jimczi to also include those, since I guess the custom score calculation could decrease performance ?

jpountz commented on server/src/main/java/org/elasticsearch/index/IndexModule.java:
```diff
@@ -144,13 +145,24 @@ public IndexModule(
+            final EngineFactory engineFactory,
+            final Map<String, IndexStorePlugin.DirectoryFactory> directoryFactories) {
+        this(indexSettings, analysisRegistry, engineFactory, directoryFactories, () -> true);
     }
```
can't we have tests call the other constructor instead and pass `()->true`?

matriv replied:
Yep, that's easy to do.

jpountz commented on server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -92,12 +93,35 @@
+                scriptService, xContentRegistry, namedWriteableRegistry, client, searcher, nowInMillis, indexNameMatcher,
+                new Index(RemoteClusterAware.buildRemoteIndexName(clusterAlias, indexSettings.getIndex().getName()),
+                        indexSettings.getIndex().getUUID()), isAllowExpensiveQueries);
+    }
```
can we avoid duplicating constructors?

matriv replied:
There are ~20 usages of this in tests, so no big deal, I can remove the constructor.

jpountz reviewed (commented):
I left some minor comments, apart from that the change looks good to me.

matriv committed:
Remove duplicate constructors to ease up tests

matriv committed:
fix tests

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

romseygeek commented on server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -192,6 +197,10 @@ public BitSetProducer bitsetFilter(Query filter) {
     }
 
+    public boolean isAllowExpensiveQueries() {
+        return isAllowExpensiveQueries.getAsBoolean();
```
Can this just be `allowExpensiveQueries()`?  Otherwise it reads very strangely.

romseygeek reviewed (approved):
One nit around naming, but LGTM otherwise

romseygeek commented on docs/reference/query-dsl.asciidoc:
```diff
@@ -25,6 +25,27 @@ or to alter their behaviour (such as the
+
+[[query-dsl-allow-expensive-queries]]
+Allow expensive queries::
+Execution of certain types of queries have usually slow performance, which can affect the cluster performance.
```
Certain types of queries will generally execute slowly due to the way they are implemented, which can affect the stability of your cluster.

romseygeek commented on docs/reference/query-dsl/prefix-query.asciidoc:
```diff
@@ -64,4 +64,12 @@ GET /_search
+
+[[prefix-query-allow-expensive-queries]]
+===== Allow expensive queries
+Prefix queries will not be executed if <<query-dsl-allow-expensive-queries, `search.allow_expensive_queries`>>
```
However, if <<index-prefixes, `index_prefixes`>> are enabled, an optimised query is built which is not considered slow, and will be executed in spite of this setting.

romseygeek reviewed (commented):
Couple of documentation changes to make it read slightly more naturally.

matriv committed:
Address comments

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

jpountz reviewed (approved):
I left a comment about the docs, LGTM otherwise.

matriv committed:
fix docs

matriv committed:
Merge remote-tracking branch 'upstream/master' into impl-29050

matriv commented:
`master` : a8b39ed842c7770bd9275958c9f747502fd9a3ea
`7.x` : dac720d7a16201f3339866352f0a601794da2d5d

javanna commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/320_disallow_queries.yml:
```diff
@@ -0,0 +1,330 @@
+      cluster.get_settings:
+        flat_settings: true
+
+  - match: {search.allow_expensive_queries: null}
```
heya @matriv I think this assertion is problematic, it works with the java runner due to the semantics of HashMap.get , but it may fail with test runners written in other languages. I was chatting to @karmi about this and the proper way to do this null check would be is_false: search.allow_expensive_queries . Would you mind please changing this?

karmi replied:
To clarify, this fails in the Go client runner, what works is `is_false: { search.allow_expensive_queries: null }`.

javanna replied:
does is_false accept values? I thought the right syntax would be `is_false: search.allow_expensive_queries`

matriv replied:
`is_false: search.allow_expensive_queries` works with Java.

karmi replied:
My bad, sorry — `is_false: search.allow_expensive_queries` is correct.

javanna commented:
heya, we were looking with @nik9000 at failing queries against runtime fields (under development in a feature branch) when expensive queries are disallowed. Looking at how the check is performed, we noticed that it is currently executed on each shard, which translates to each shard returning the same error (which then get de-duplicated in the coordinating node). Shall we move these checks to the coordinating node? Ideally, expensive queries would be rejected straight-away before being sent to the shards as part of the query phase.

matriv commented:
@javanna This sounds good to me.
@jimczi Do you maybe have something to add on this proposed change/improvement?

matriv commented:
@javana Jim reminded me of the reason we didn't do it in the first place: Currently the mapping is resolved on every shard and for example: a prefix query on a field with some options enabled can be turned into a simple term query so with the proposed approach we would return an error on queries that we accept today (because of the rewrite simplification) when the option is enabled.

Maybe, in the future we could enhance this behaviour by resolving the mapping the co-ordinating node, possibly using `field caps`.
